File: src/org/jgll/util/Input.java
package org.jgll.util;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

import org.jgll.sppf.SPPFNode;
import org.jgll.traversal.PositionInfo;

/**
 * 
 * Is backed by an integer array.
 * 
 * @author Ali Afroozeh
 *
 */
public class Input {

	private int[] input;

	private LineColumn[] lineColumns;
	
	public static Input fromString(String s) {
		int[] input = new int[s.length() + 1];
		for (int i = 0; i < s.length(); i++) {
			input[i] = s.codePointAt(i);
		}
		// The EOF character is assumed to have value 0 instead of the more common -1.  
		// as Bitsets cannot work with negative values. 
		input[s.length()] = 0;

		return new Input(input);
	}
	
	public static Input fromIntArray(int[] input) {
		return new Input(input);
	}
	
	public static Input fromPath(String path) throws IOException {
		return fromString(readTextFromFile(path));
	}

	private Input(int[] input) {
		this.input = input;
		lineColumns = new LineColumn[input.length];
		calculateLineLengths();
	}
	
	public int charAt(int index) {
		return input[index];
	}

	public int size() {
		return input.length;
	}
	
	/**
	 * Returns the whole contents of a text file as a string.
	 * 
	 * @param path
	 *            the path to the text file
	 * @throws IOException
	 */
	private static String readTextFromFile(String path) throws IOException {
		return readTextFromFile(new File(path));
	}

	private static String readTextFromFile(File file) throws IOException {
		StringBuilder sb = new StringBuilder();

		InputStream in = new BufferedInputStream(new FileInputStream(file));
		int c = 0;
		while ((c = in.read()) != -1) {
			sb.append((char) c);
		}

		in.close();

		return sb.toString();
	}

	public static String read(InputStream is) throws IOException {
		StringBuilder sb = new StringBuilder();
		BufferedInputStream in = new BufferedInputStream(is);
		int c = 0;
		while ((c = in.read()) != -1) {
			sb.append((char) c);
		}
		in.close();
		return sb.toString();
	}
	
	public int[] subInput(int start, int end) {
		int[] subInput = new int[end - start + 1];
		
		System.arraycopy(input, start, subInput, start, end - start + 1);
		
		return subInput;
	}
	
	public boolean match(int start, int end, String target) {
		return match(start, end, toIntArray(target));
	}
	
	public boolean match(int start, int end, int[] target) {
		if(target.length != end - start) {
			return false;
		}
	 	
		int i = 0;
		while(i < target.length) {
			if(target[i] != input[start + i]) {
				return false;
			}
			i++;
		}
		
		return true;
	}

	public boolean match(int from, String target) {
		return match(from, toIntArray(target));
	}
	
	public boolean matchBackward(int start, String target) {
		return matchBackward(start, toIntArray(target));
	}
	
	public boolean matchBackward(int start, int[] target) {
		if(start - target.length < 0) {
			return false;
		}
		
		int i = target.length - 1;
		int j = start - 1;
		while(i >= 0) {
			if(target[i] != input[j]) {
				return false;
			}
			i--;
			j--;
		}
		
		return true;
	}
	
	public boolean match(int from, int[] target) {
		
		if(target.length > size() - from) {
			return false;
		}
		
		int i = 0;
		while(i < target.length) {
			if(target[i] != input[from + i]) {
				return false;
			}
			i++;
		}
		
		return true;
	}
	
	public static int[] toIntArray(String s) {
		int[] array = new int[s.codePointCount(0, s.length())];
		for(int i = 0; i < array.length; i++) {
			array[i] = s.codePointAt(i);
		}
		return array;
	}
 
	public int getLineNumber(int index) {
		if(index < 0) {
			return 0;
		}
		return lineColumns[index].getLineNumber();
	}
	
	public int getColumnNumber(int index) {
		if(index < 0) {
			return 0;
		}
		return lineColumns[index].getColumnNumber();
	}
	
	public PositionInfo getPositionInfo(SPPFNode node) {
		return new PositionInfo(node.getLeftExtent(), 
				node.getRightExtent() - node.getLeftExtent(), 
				getLineNumber(node.getLeftExtent()), 
				getColumnNumber(node.getLeftExtent()), 
				getLineNumber(node.getRightExtent()), 
				getColumnNumber(node.getRightExtent()));
	}

	private void calculateLineLengths() {
		int lineNumber = 1;
		int columnNumber = 1;

		// Empty input: only the end of line symbol
		if(input.length == 1) {
			lineColumns[0] = new LineColumn(lineNumber, columnNumber);
			return;
		}
		
		for (int i = 0; i < input.length - 1; i++) {
			lineColumns[i] = new LineColumn(lineNumber, columnNumber);
			if (input[i] == '\n') {
				lineNumber++;
				columnNumber = 1;
			} else if(input[i] == '\r' && i < input.length - 2 && input[i + 1] == '\n') {
				columnNumber = 1;
				lineNumber++;
				i++;
			} else if (input[i] == '\r') {
				lineNumber++;
				columnNumber = 1;
			} else {
				columnNumber++;
			}
		}
		
		// The end of the line char column as the last character
		lineColumns[input.length - 1] = new LineColumn(lineColumns[input.length - 2]);
	}
		
	private static class LineColumn {

		private int lineNumber;
		private int columnNumber;
		
		public LineColumn(int lineNumber, int columnNumber) {
			this.lineNumber = lineNumber;
			this.columnNumber = columnNumber;
		}
		
		public LineColumn(LineColumn lineColumn) {
			this.lineNumber = lineColumn.lineNumber;
			this.columnNumber = lineColumn.columnNumber; 
		}
		
		public int getLineNumber() {
			return lineNumber;
		}
		
		public int getColumnNumber() {
			return columnNumber;
		}
		
		@Override
		public String toString() {
			return "(" + lineNumber + ":" + columnNumber + ")";
		}
		
		@Override
		public boolean equals(Object obj) {
			if(this == obj) {
				return true;
			}
			
			if(!(obj instanceof LineColumn)) {
				return false;
			}
			
			LineColumn other = (LineColumn) obj;
			return lineNumber == other.lineNumber && columnNumber == other.columnNumber;
		}
	}
	
	@Override
	public String toString() {
		
		List<Character> charList = new ArrayList<>();
		for(int i : input) {
			char[] chars = Character.toChars(i);
			for(char c : chars) {
				charList.add(c);
			}
		}
		
		StringBuilder sb = new StringBuilder();
		sb.append("[");
		for(char c : charList) {
			sb.append(c).append(", ");
		}
		sb.append("]");
		
		return sb.toString();
	}
}

File: src-scala/org/jgll_staged/util/Input.scala
package org.jgll_staged.util

import java.io.BufferedInputStream
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream
import java.util.ArrayList
import java.util.Arrays
import java.util.List
import org.jgll_staged.traversal.PositionInfo
import Input._
import scala.reflect.{BeanProperty, BooleanBeanProperty}
//remove if not needed
import scala.collection.JavaConversions._

object Input {

  def fromString(s: String): Input = {
    val input = Array.ofDim[Int](s.length + 1)
    for (i <- 0 until s.length) {
      input(i) = s.codePointAt(i)
    }
    input(s.length) = 0
    new Input(input)
  }

  def fromIntArray(input: Array[Int]): Input = new Input(input)

  def fromPath(path: String): Input = fromString(readTextFromFile(path))

  private def readTextFromFile(path: String): String = readTextFromFile(new File(path))

  private def readTextFromFile(file: File): String = {
    val sb = new StringBuilder()
    val in = new BufferedInputStream(new FileInputStream(file))
    var c = 0
    def setC(i: Int) = { c = i; c }

    while (setC(in.read()) != -1) {
      sb.append(c.toChar)
    }
    in.close()
    sb.toString
  }

  def read(is: InputStream): String = {
    val sb = new StringBuilder()
    val in = new BufferedInputStream(is)
    var c = 0
    def setC(i: Int) = { c = i; c }

    while (setC(in.read()) != -1) {
      sb.append(c.toChar)
    }
    in.close()
    sb.toString
  }

  def toIntArray(s: String): Array[Int] = {
    val array = Array.ofDim[Int](s.codePointCount(0, s.length))
    for (i <- 0 until array.length) {
      array(i) = s.codePointAt(i)
    }
    array
  }

  private class LineColumn(@BeanProperty var lineNumber: Int, @BeanProperty var columnNumber: Int)
      {

    def this(lineColumn: LineColumn) {
      this(lineColumn.lineNumber, lineColumn.columnNumber)
    }

    override def toString(): String = {
      "(" + lineNumber + ":" + columnNumber + ")"
    }

    override def equals(obj: Any): Boolean = {
      if (this == obj) {
        return true
      }
      if (!(obj.isInstanceOf[LineColumn])) {
        return false
      }
      val other = obj.asInstanceOf[LineColumn]
      lineNumber == other.lineNumber && columnNumber == other.columnNumber
    }
  }
}

class Input private (private var input: Array[Int]) {

  private var lineColumns: Array[LineColumn] = new Array[LineColumn](input.length)

  calculateLineLengths()

  def charAt(index: Int): Int = input(index)

  def size(): Int = input.length

  def subInput(start: Int, end: Int): Array[Int] = {
    val length = end - start + 1
    val subInput = Array.ofDim[Int](length)
    System.arraycopy(input, start, subInput, 0, length)
    subInput
  }

  def `match`(start: Int, end: Int, target: String): Boolean = `match`(start, end, toIntArray(target))

  def `match`(start: Int, end: Int, target: Array[Int]): Boolean = {
    if (target.length != end - start) {
      return false
    }
    var i = 0
    while (i < target.length) {
      if (target(i) != input(start + i)) {
        return false
      }
      i += 1
    }
    true
  }

  def `match`(from: Int, target: String): Boolean = `match`(from, toIntArray(target))

  def matchBackward(start: Int, target: String): Boolean = {
    matchBackward(start, toIntArray(target))
  }

  def matchBackward(start: Int, target: Array[Int]): Boolean = {
    if (start - target.length < 0) {
      return false
    }
    var i = target.length - 1
    var j = start - 1
    while (i >= 0) {
      if (target(i) != input(j)) {
        return false
      }
      i -= 1
      j -= 1
    }
    true
  }

  def `match`(from: Int, target: Array[Int]): Boolean = {
    if (target.length > size - from) {
      return false
    }
    var i = 0
    while (i < target.length) {
      if (target(i) != input(from + i)) {
        return false
      }
      i += 1
    }
    true
  }

  def getLineNumber(index: Int): Int = {
    if (index < 0 || index >= lineColumns.length) {
      return 0
    }
    lineColumns(index).getLineNumber
  }

  def getColumnNumber(index: Int): Int = {
    if (index < 0 || index >= lineColumns.length) {
      return 0
    }
    lineColumns(index).getColumnNumber
  }

  override def equals(obj: Any): Boolean = {
    if (this == obj) {
      return true
    }
    if (!(obj.isInstanceOf[Input])) {
      return false
    }
    val other = obj.asInstanceOf[Input]
    Arrays.equals(input, other.input)
  }

  def getPositionInfo(leftExtent: Int, rightExtent: Int): PositionInfo = {
    new PositionInfo(leftExtent, rightExtent - leftExtent, getLineNumber(leftExtent), getColumnNumber(leftExtent), 
      getLineNumber(rightExtent), getColumnNumber(rightExtent))
  }

  private def calculateLineLengths() {
    var lineNumber = 1
    var columnNumber = 1
    if (input.length == 1) {
      lineColumns(0) = new LineColumn(lineNumber, columnNumber)
      return
    }
    for (i <- 0 until input.length - 1) {
      lineColumns(i) = new LineColumn(lineNumber, columnNumber)
      if (input(i) == '\n') {
        lineNumber += 1
        columnNumber = 1
      } else if (input(i) == '\r') {
        columnNumber = 1
      } else {
        columnNumber += 1
      }
    }
    lineColumns(input.length - 1) = new LineColumn(lineColumns(input.length - 2))
  }

  override def toString(): String = {
    val charList = new ArrayList[Character]()
    for (i <- input) {
      val chars = Character.toChars(i)
      for (c <- chars) {
        charList.add(c)
      }
    }
    val sb = new StringBuilder()
    for (c <- charList) {
      sb.append(c)
    }
    sb.toString
  }

  def isEndOfLine(currentInputIndex: Int): Boolean = {
    input(currentInputIndex) == 0 || 
      lineColumns(currentInputIndex + 1).columnNumber == 0
  }

  def isStartOfLine(currentInputIndex: Int): Boolean = {
    currentInputIndex == 0 || lineColumns(currentInputIndex).columnNumber == 0
  }
}
